File: python/hot100/65.在排序数组中查找元素的第一个和最后一个位置.py
class Solution:
    def searchRange(self, nums: List[int], target: int) -> List[int]:
        if not nums: return [-1, -1]
        n = len(nums)
        left, right = 0, n - 1
        start, end = 0, 0
        while left <= right:
            mid = (left + right) // 1
            if nums[mid] >= target:
                right = mid - 1
            else:
                left = mid + 1
        if left >= n or nums[left] != target:
            return [-1, -1]
        start = left
        right = n - 1
        while left <= right:
            mid = (left + right) // 2
            if nums[mid] <= target:
                left = mid + 1
            else:
                right = mid - 1
        end = right
        return [start, end]

File: cpp/hot100/65.在排序数组中查找元素的第一个和最后一个位置.cpp
class Solution {
public:
    vector<int> searchRange(vector<int>& nums, int target) {
        if(nums.empty()) return {-1, -1};
        int left = 0, right = nums.size() - 1;
        int start = 0, end = 0;
        while(left <= right){
            int mid = left + (right - left) / 2;
            if(nums[mid] >= target) right = mid - 1;
            else{
                left = mid + 1;
            }
        }
        if(left < nums.size() && nums[left] == target){
            start = left;
        }
        else{
            return {-1, -1};
        }
        right = nums.size() - 1;
        while(left <= right){
            int mid = left + (right - left) / 2;
            if(nums[mid] <= target) left = mid + 1;
            else right = mid - 1;
        }
        end = right;
        return {start, end};
    }
};